src/python/library/tritonclient/utils/cuda_shared_memory/__init__.py

```diff
@@ -258,10 +299,66 @@ def get_contents_as_numpy(cuda_shm_handle, datatype, shape):
             result = np.reshape(val, shape)
 
     finally:
-        c_int(_cshm_cuda_shared_memory_release_buffer(shm_addr))
+        c_int(_cshm_cuda_shared_memory_release_host_buffer(host_addr))
         return result
 
 
+def set_shared_memory_region_from_dlpack(cuda_shm_handle, input_values):
+    # this function basically is an implementation of 'from_dlpack'
+    offset_current = 0
+    for input_value in input_values:
+        # Knowing the implementation detail of how shared memory region is
+        # set (cudaMemcpy). There is no need to transfer ownership of
+        # 'dl_managed_tensor': the data has been copied out when dlpack
+        # capsule is out of scope.
+        stream = _dlpack_stream.value if bool(_dlpack_stream) else None
+        dlcapsule = _dlpack.get_dlpack_capsule(input_value, stream)
+
+        ptr = ctypes.pythonapi.PyCapsule_GetPointer(dlcapsule,
+                                                    _dlpack.c_str_dltensor)
+        dmt = _dlpack.DLManagedTensor.from_address(ptr)
+        if not _dlpack.is_contiguous_data(
+                dmt.dl_tensor.ndim, dmt.dl_tensor.shape, dmt.dl_tensor.strides):
+            _raise_error(
+                "DLPack tensor is not contiguous. Only contiguous DLPack tensors that are stored in C-Order are supported."
+            )
+
+        # Write to shared memory region
+        byte_size = _dlpack.get_byte_size(dmt.dl_tensor.dtype,
+                                          dmt.dl_tensor.ndim,
+                                          dmt.dl_tensor.shape)
+        # apply offset to the data pointer ('data' pointer is implicitly converted to int)
+        data_ptr = dmt.dl_tensor.data + dmt.dl_tensor.byte_offset
+        if dmt.dl_tensor.device == _dlpack.DLDeviceType.kDLCUDA:
+            device_id = dmt.dl_tensor.device.device_id
+        else:
+            device_id = -1
+
+        _raise_if_error(c_int(_ccudashm_stream_synchronize(_dlpack_stream)))
+        _raise_if_error(
+            c_int(_ccudashm_shared_memory_region_set(cuda_shm_handle, c_uint64(offset_current), \
+                c_uint64(byte_size), cast(data_ptr, c_void_p), device_id)))
```

Comment:
I think we need a CUDA synchronization here too since it is a D2D copy and even using cudaMemcpy would not be synchronous.

Reply:
`cudaMemcpy` doesn't accept `stream` argument, so a global synchronization needs to be performed? If so, we should have use `cudaMemcpyAsync` to minimize the scope of synchronization, and I will do it as part of the Python-CUDA replacement.